src/Microsoft.OData.Core/ResourceSetWithoutExpectedTypeValidator.cs

```diff
@@ -71,6 +71,12 @@ internal void ValidateResource(IEdmType itemType)
                 throw new ODataException(Strings.ResourceSetWithoutExpectedTypeValidator_IncompatibleTypes(itemType.FullTypeName(), this.itemType.FullTypeName()));
             }
 
+            // Make sure the types are in the same inheritance tree
+            if (!(itemType.IsAssignableFrom(this.itemType) || this.itemType.IsAssignableFrom(itemType)))

```

Comment:
>| this.itemType.IsAssignableFrom(itemType) [](start = 60, length = 42)

This would be correct if the expected type was unknown but it seems like this validator is always constructed with an IEdmType. Is the class wrongly named? May be we should check this only when no item type is provided in the constructor.

Reply:
I think there are two cases:
1) We were provided an expected type, and we want to validate that the type is assignable from the expected type (and not the other way around)
2) We "guessed" the expected type, based on the type of the first item in the collection. In this case, we should support types in the same type hierarchy (although, technically, with the introduction of the abstract base types Edm.Primitive, Edm.ComplexType, Edm.EntityType, and Edm.Untyped, even this validation may not be valid). So, we might just want to remove the logic that guesses at the type based on the type of the first element.

---
In reply to: [283918035](https://github.com/OData/odata.net/pull/1463#discussion_r283918035) [](ancestors = 283918035)